File: old/actions/fix_apt_sources.py
"""
Fix apt sources action
"""
import logging
from .base import Action
from cli import Sed
logger = logging.getLogger(__name__)

class FixAptSourcesAction(Action):
    """Fix apt sources (replace oracular with plucky, old-releases with archive)"""
    description = "apt sources fix"

    def execute(self) -> bool:
        """Fix apt sources"""
        sed_cmds = [
            Sed().replace("/etc/apt/sources.list", "oracular", "plucky"),
            Sed().replace("/etc/apt/sources.list", "old-releases.ubuntu.com", "archive.ubuntu.com")
        ]
        all_succeeded = True
        for sed_cmd in sed_cmds:
            output, exit_code = self.pct_service.execute(self.container_id, sed_cmd)
            if exit_code is not None and exit_code != 0:
                # If file doesn't exist, that's OK (Ubuntu 25.04 might use sources.list.d)
                if "No such file or directory" in output or "can't read" in output:
                    logger.info("sources.list not found (may use sources.list.d), skipping fix")
                else:
                    logger.error("Failed to fix apt sources: %s", output)
                    all_succeeded = False
        return all_succeeded

File: src/actions/fix_apt_sources.go
package actions

import (
	"strconv"
	"enva/cli"
	"enva/libs"
	"enva/services"
	"strings"
)

// FixAptSourcesAction fixes apt sources
type FixAptSourcesAction struct {
	*BaseAction
}

func NewFixAptSourcesAction(sshService *services.SSHService, aptService *services.APTService, pctService *services.PCTService, containerID *string, cfg *libs.LabConfig, containerCfg *libs.ContainerConfig) Action {
	return &FixAptSourcesAction{
		BaseAction: &BaseAction{
			SSHService:   sshService,
			APTService:   aptService,
			PCTService:   pctService,
			ContainerID: containerID,
			Cfg:          cfg,
			ContainerCfg: containerCfg,
		},
	}
}

func (a *FixAptSourcesAction) Description() string {
	return "apt sources fix"
}

func (a *FixAptSourcesAction) Execute() bool {
	sedCmds := []string{
		cli.NewSed().Replace("/etc/apt/sources.list", "oracular", "plucky"),
		cli.NewSed().Replace("/etc/apt/sources.list", "old-releases.ubuntu.com", "archive.ubuntu.com"),
	}
	containerIDInt, err := strconv.Atoi(*a.ContainerID)
	if err != nil {
		libs.GetLogger("fix_apt_sources").Printf("Invalid container ID: %s", *a.ContainerID)
		return false
	}
	allSucceeded := true
	for _, sedCmd := range sedCmds {
		output, exitCode := a.PCTService.Execute(containerIDInt, sedCmd, nil)
		if exitCode != nil && *exitCode != 0 {
			if strings.Contains(output, "No such file or directory") || strings.Contains(output, "can't read") {
				libs.GetLogger("fix_apt_sources").Printf("sources.list not found (may use sources.list.d), skipping fix")
			} else {
				libs.GetLogger("fix_apt_sources").Printf("Failed to fix apt sources: %s", output)
				allSucceeded = false
			}
		}
	}
	return allSucceeded
}

